libraries/abstractions/pkcs11/psa/iot_pkcs11_psa_object_management.c

```diff
@@ -0,0 +1,600 @@
+								   ulDataSize, pucData, uxFlags );
+
+			if( uxStatus == PSA_PS_SUCCESS )
+			{
+				xHandle = eAwsDeviceCertificate;
+
+				/*change sst written mark'*/
+				P11KeyConfig.ulDeviceCertificateMark = ulPresentMark;
+			}
+		}
+
+		/*
+		 * write Jitp certificate.
+		 */
+		else if( ( *( ( uint32_t * )pxClass->pValue ) == CKO_CERTIFICATE ) &&
+			( strcmp( pxLabel->pValue,
+					pkcs11configLABEL_JITP_CERTIFICATE ) == 0 ) )
+		{
+			uxStatus = psa_ps_set( P11KeyConfig.uxJitpCertificate,
+								   ulDataSize, pucData, uxFlags);
+
+			if( uxStatus == PSA_PS_SUCCESS )
+			{
+				xHandle = eAwsDeviceCertificate;
+
+				/*change sst written mark'*/
+				P11KeyConfig.ulJitpCertificateMark = ulPresentMark;
+			}
+		}
+
+		/*
+		 * write Root certificate.
+		 */
+		else if( ( *( ( uint32_t * )pxClass->pValue ) == CKO_CERTIFICATE ) &&
+			( strcmp( pxLabel->pValue,
+					pkcs11configLABEL_ROOT_CERTIFICATE ) == 0 ) )
+		{
+			uxStatus = psa_ps_set( P11KeyConfig.uxRootCertificate,
+								   ulDataSize, pucData, uxFlags);
+
+			if( uxStatus == PSA_PS_SUCCESS )
+			{
+				xHandle = eAwsDeviceCertificate;
+
+				/*change sst written mark'*/
+				P11KeyConfig.ulRootCertificateMark = ulPresentMark;
+			}
+		}
+
+		/*
+		 * Store device private key by crypto service of PSA.
+		 */
+
+		else if( ( *( ( uint32_t * )pxClass->pValue ) == CKO_PRIVATE_KEY ) &&
+			( strcmp( pxLabel->pValue,
+					pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS ) == 0 ) )
+		{
+			/*
+			 * Currently TF-M does not support persistent key APIs, so volatile
+			 * key is used here.
+			 */
+			/*
+			 * TODO: Replace the volatile key APIs with persistent key APIs
+			 * after TF-M support that.
+			 */
+			uxStatus = psa_allocate_key( &uxKeyHandle );
+
+			/*
+			 * Device private key is only used to make a signature,
+			 * only support RSA alg, only support MBEDTLS_RSA_PKCS_V15
+			 * padding mode, only support SHA256 md alg.
+			 */
+			if ( uxStatus == PSA_SUCCESS )
+			{
+				uxPrivateKeyTypePKCS11 = mbedtls_pk_get_type( pvContext );
+				switch ( uxPrivateKeyTypePKCS11 )
+				{
+					case MBEDTLS_PK_RSA:
+						uxKeyType = PSA_KEY_TYPE_RSA_KEYPAIR;
+						pucKeyData = pucData;
+						ulKeyDataSize = ulDataSize;
+						switch ( ( ( mbedtls_rsa_context * ) ( pvContext->pk_ctx ) )->padding )
+						{
+							case MBEDTLS_RSA_PKCS_V15:
+								uxAlgorithm = PSA_ALG_RSA_PKCS1V15_SIGN( PSA_ALG_SHA_256 );
+								break;
+
+							case MBEDTLS_RSA_PKCS_V21:
+								uxAlgorithm = PSA_ALG_RSA_PSS( PSA_ALG_SHA_256 );
+								break;
+
+							default:
+								uxAlgorithm = 0;
+								break;
+						}
+						break;
+
+					case MBEDTLS_PK_ECKEY:
+					case MBEDTLS_PK_ECDSA:
+						ec = (mbedtls_ecp_keypair *) (pvContext->pk_ctx );
+						curve_id = mbedtls_ecp_curve_info_from_grp_id( ec->grp.id )->tls_id;
+						uxKeyType = PSA_KEY_TYPE_ECC_KEYPAIR(curve_id);
+						uxAlgorithm = PSA_ALG_ECDSA( PSA_ALG_SHA_256 );
+						if ( 0 != get_ECPrivateKey_sequence_privatekey_sec1_der(
```

Comment:
With DER format, a leading zero will be added if the MSB of key is "1",  the key size will be one byte more than the algorithm key size. In this case, the key will be rejected the key length doesn't match. Contacted mbed-crypto developer,  I was told only raw (binary) format of private key is supported for psa_import_key().

Here is my change for that
                                                ec = mbedtls_pk_ec(*pvContext);;
                                                curve_id = mbedtls_ecp_curve_info_from_grp_id( ec->grp.id )->tls_id;
                                                uxKeyType = PSA_KEY_TYPE_ECC_KEYPAIR(curve_id);
                                                uxAlgorithm = PSA_ALG_ECDSA( PSA_ALG_SHA_256 );
                                                xPrivateKeySizeRaw = ( ec->grp.nbits + 7 ) / 8;
                                                if( mbedtls_mpi_write_binary( &ec->d, pcPrivateKeyRaw, xPrivateKeySizeRaw ) != 0 )
                                                {
                                                        uxStatus = PSA_ERROR_GENERIC_ERROR;
                                                }
                                                pucKeyData = pcPrivateKeyRaw;
                                                ulKeyDataSize = xPrivateKeySizeRaw;

Reply:
First of all, thanks for your propose here.
I am sorry that the name of the function get_ECPrivateKey_sequence_privatekey_sec1_der is a lit confusing. Let me explain the solution I used. This function is to extract the privateKey and its length from the key in DER format. When encoding the key into DER binary format, the privateKey field is encoded by calling the  mbedtls_asn1_write_mpi function in which the length of the privateKey field is calculated(as you described, ( ( mbedtls_mpi_bitlen( ec->d ) + 7 ) >> 3 ) ) and the data length of stored in the address next to the privateKey encoded data. The data length, as well as the privateKey field, is encoded as the MBEDTLS_ASN1_OCTET_STRING tag. The function get_ECPrivateKey_sequence_privatekey_sec1_der is to extract the privateKey field binary data and its length. It looks for the address where the privateKey(as well as its length) is stored and then get the length of the tag MBEDTLS_ASN1_OCTET_STRING which is exactly the privateKey length. Thus we get the start address and the length of the privateKey binary format which is exactly what we need.
So your proposal is very similar to mine except for the data length input.
The input data length should be the binary format privateKey's data length. The psa_import_key uses the data start address and its length to check whether the input data is valid. So I think it isn't reasonable the take ( ec->grp.nbits + 7 ) / 8 as the input data length. Input data length should be the length of what your really input anyway.

Reply:
With your original code, we see error while running AFR IDT test. the private key is rejected as key size mis-match (a leading zero in DER format, resulting in the key size is 33 bytes, instead of 32 bytes). I spoken to mbedcrypto developer, they say PSA SST only support raw format now. So I made the change to convert the key in raw format.

Reply:
yes, you are right. The raw binary format of privateKey should be taken as the input parameter. I have updated the code to fix it.